File: src/main/java/com/smpcore/sidebar/SidebarManager.java
package com.smpcore.sidebar;

import com.smpcore.config.ConfigManager;
import com.smpcore.config.SmpSettings;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitTask;
import org.bukkit.scoreboard.DisplaySlot;
import org.bukkit.scoreboard.Objective;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public final class SidebarManager {
    private static final String OBJECTIVE = "smpcore_status";

    private final JavaPlugin plugin;
    private final ConfigManager configManager;
    private final Map<UUID, Scoreboard> scoreboards = new HashMap<>();
    private BukkitTask refreshTask;

    public SidebarManager(JavaPlugin plugin, ConfigManager configManager) {
        this.plugin = plugin;
        this.configManager = configManager;
    }

    public void start() {
        stop();
        SmpSettings settings = configManager.settings();
        if (!settings.sidebarEnabled()) {
            clearAll();
            return;
        }

        refreshAll();
        this.refreshTask = Bukkit.getScheduler().runTaskTimer(
                plugin,
                this::refreshAll,
                settings.sidebarUpdateIntervalTicks(),
                settings.sidebarUpdateIntervalTicks()
        );
    }

    public void stop() {
        if (refreshTask != null) {
            refreshTask.cancel();
            refreshTask = null;
        }
    }

    public void onJoin(Player player) {
        if (!configManager.settings().sidebarEnabled()) {
            return;
        }
        if (!player.hasPermission("smpcore.sidebar.view")) {
            return;
        }
        applySidebar(player, configManager.settings());
    }

    public void onQuit(Player player) {
        scoreboards.remove(player.getUniqueId());
    }

    public void refreshAll() {
        SmpSettings settings = configManager.settings();
        if (!settings.sidebarEnabled()) {
            clearAll();
            return;
        }

        for (Player player : Bukkit.getOnlinePlayers()) {
            if (!player.hasPermission("smpcore.sidebar.view")) {
                continue;
            }
            applySidebar(player, settings);
        }
    }

    public void clearAll() {
        for (Player player : Bukkit.getOnlinePlayers()) {
            player.setScoreboard(Bukkit.getScoreboardManager().getMainScoreboard());
        }
        scoreboards.clear();
    }

    private void applySidebar(Player player, SmpSettings settings) {
        Scoreboard scoreboard = scoreboards.computeIfAbsent(
                player.getUniqueId(),
                ignored -> Bukkit.getScoreboardManager().getNewScoreboard()
        );

        Objective objective = scoreboard.getObjective(OBJECTIVE);
        if (objective == null) {
            objective = scoreboard.registerNewObjective(OBJECTIVE, "dummy", color(settings.sidebarTitle()));
            objective.setDisplaySlot(DisplaySlot.SIDEBAR);
        } else {
            objective.setDisplayName(color(settings.sidebarTitle()));
        }

        reset(scoreboard);

        List<String> lines = settings.sidebarLines();
        int score = lines.size();
        for (int i = 0; i < lines.size(); i++) {
            String entry = ChatColor.values()[i].toString() + ChatColor.RESET;
            Team team = scoreboard.getTeam("line_" + i);
            if (team == null) {
                team = scoreboard.registerNewTeam("line_" + i);
                team.addEntry(entry);
            }
            team.prefix(color(lines.get(i)));
            objective.getScore(entry).setScore(score--);
        }

        player.setScoreboard(scoreboard);
    }

    private void reset(Scoreboard scoreboard) {
        for (String entry : scoreboard.getEntries()) {
            scoreboard.resetScores(entry);
        }
    }

    private String color(String text) {
        return ChatColor.translateAlternateColorCodes('&', text);
    }
}

File: src/main/kotlin/com/smpcore/sidebar/SidebarManager.kt
package com.smpcore.sidebar

import com.smpcore.config.ConfigManager
import com.smpcore.config.SmpSettings
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitTask
import org.bukkit.scoreboard.Criteria
import org.bukkit.scoreboard.DisplaySlot
import org.bukkit.scoreboard.Scoreboard
import java.util.UUID

class SidebarManager(private val plugin: JavaPlugin, private val configManager: ConfigManager) {
    private val objectiveName = "smpcore_status"
    private val scoreboards = mutableMapOf<UUID, Scoreboard>()
    private var refreshTask: BukkitTask? = null

    fun start() {
        stop()
        val settings = configManager.settings()
        if (!settings.sidebarEnabled) {
            clearAll(); return
        }
        refreshAll()
        refreshTask = Bukkit.getScheduler().runTaskTimer(
            plugin,
            Runnable { refreshAll() },
            settings.sidebarUpdateIntervalTicks,
            settings.sidebarUpdateIntervalTicks
        )
    }

    fun stop() {
        refreshTask?.cancel()
        refreshTask = null
    }

    fun onJoin(player: Player) {
        if (!configManager.settings().sidebarEnabled) return
        if (!player.hasPermission("smpcore.sidebar.view")) return
        applySidebar(player, configManager.settings())
    }

    fun onQuit(player: Player) {
        scoreboards.remove(player.uniqueId)
    }

    fun refreshAll() {
        val settings = configManager.settings()
        if (!settings.sidebarEnabled) {
            clearAll(); return
        }

        Bukkit.getOnlinePlayers().forEach { p ->
            if (p.hasPermission("smpcore.sidebar.view")) applySidebar(p, settings)
        }
    }

    fun clearAll() {
        val main = Bukkit.getScoreboardManager().mainScoreboard
        Bukkit.getOnlinePlayers().forEach { it.scoreboard = main }
        scoreboards.clear()
    }

    private fun applySidebar(player: Player, settings: SmpSettings) {
        val manager = Bukkit.getScoreboardManager()
        val scoreboard = scoreboards.computeIfAbsent(player.uniqueId) { manager.newScoreboard }

        var objective = scoreboard.getObjective(objectiveName)
        if (objective == null) {
            objective = scoreboard.registerNewObjective(objectiveName, Criteria.DUMMY, color(settings.sidebarTitle))
            objective.displaySlot = DisplaySlot.SIDEBAR
        } else {
            objective.displayName = color(settings.sidebarTitle)
        }

        scoreboard.entries.forEach { scoreboard.resetScores(it) }

        var score = settings.sidebarLines().size
        settings.sidebarLines().forEachIndexed { index, line ->
            val entry = ChatColor.values()[index].toString() + ChatColor.RESET
            var team = scoreboard.getTeam("line_$index")
            if (team == null) {
                team = scoreboard.registerNewTeam("line_$index")
                team.addEntry(entry)
            }
            team.prefix(color(line))
            objective.getScore(entry).score = score--
        }

        player.scoreboard = scoreboard
    }

    private fun color(text: String): String = ChatColor.translateAlternateColorCodes('&', text)
}
